File: backend/app/services/prime_service.py
"""
prime_service : gere le cycle de vie d'une Prime (lecture, filtres) et
delegue tout calcul au prime_calculation_service, conformement a la
separation des responsabilites definie dans l'architecture technique.
"""
from sqlalchemy.orm import Session

from app.models.prime import Prime
from app.models.pos import POS


def list_primes(db: Session, partner_id: int, period_id: int | None = None,
                 status: str | None = None, skip: int = 0, limit: int = 100) -> dict:
    query = db.query(Prime).join(POS).filter(POS.partner_id == partner_id)
    if period_id:
        query = query.filter(Prime.prime_period_id == period_id)
    if status:
        query = query.filter(Prime.status == status)
    total = query.order_by(None).count()
    items = query.offset(skip).limit(limit).all()

    # Champs d'affichage pour les tableaux frontend (code/nom POS, partenaire,
    # periode) : attributs transitoires lus par PrimeOut via from_attributes.
    for prime in items:
        pos = prime.pos
        prime.pos_code = pos.code_pos if pos else None
        prime.pos_nom = pos.name if pos else None
        prime.partner_name = pos.partner.name if pos and pos.partner else None
        prime.period_code = prime.prime_period.code if prime.prime_period else None

    return {"items": items, "total": total, "skip": skip, "limit": limit,
            "has_next": skip + len(items) < total}

File: frontend/src/services/primeService.js
import api from './api';

const normalizeList = (data) => {
  if (Array.isArray(data)) return data;
  if (Array.isArray(data?.items)) return data.items;
  if (Array.isArray(data?.data)) return data.data;
  if (Array.isArray(data?.results)) return data.results;
  return [];
};

const normalizePrime = (prime) => ({
  ...prime,
  id: prime?.id,
  montant: prime?.montant ?? prime?.amount ?? 0,
  date_attribution: prime?.date_attribution ?? prime?.date ?? '',
  statut: prime?.statut ?? prime?.status ?? '',
  pos: prime?.pos ? {
    ...prime.pos,
    nom: prime.pos.nom ?? prime.pos.name ?? '',
    name: prime.pos.name ?? prime.pos.nom ?? '',
    code_pos: prime.pos.code_pos ?? prime.pos.code ?? '',
    partenaire: prime.pos.partenaire ? {
      ...prime.pos.partenaire,
      nom: prime.pos.partenaire.nom ?? prime.pos.partenaire.name ?? '',
      name: prime.pos.partenaire.name ?? prime.pos.partenaire.nom ?? '',
    } : prime.pos.partenaire,
  } : prime?.pos,
  partenaire: prime?.partenaire ? {
    ...prime.partenaire,
    nom: prime.partenaire.nom ?? prime.partenaire.name ?? '',
    name: prime.partenaire.name ?? prime.partenaire.nom ?? '',
  } : prime?.partenaire,
});

export const primeService = {
  getAll: async (params) => {
    const response = await api.get('/primes', { params });
    const list = normalizeList(response.data).map(normalizePrime);
    return { ...response, data: { ...(response.data || {}), items: list } };
  },
};

export default primeService;
